azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java

```diff
@@ -142,6 +146,45 @@ private MySQLBasicDataSource(String host, int port, String dbName,
       setTestOnBorrow(true);
     }
 
+    /**
+     * This method overrides {@link BasicDataSource#getConnection()}, in order to have retry logics.
+     *
+     */
+    @Override
+    public Connection getConnection() throws SQLException {
+
+      /*
+       * when az server process launches, if the connection isn't valid (e.g., network problem, wrong password, etc),
+       * we fail the connection immediately.
+       */
+      if (getDataSource() == null) {
+        return createDataSource().getConnection();
+      }
+
+      Connection connection = null;
+      int retryAttempt = 0;
+      this.dataSource = null;
+      while (connection == null && retryAttempt < ServerInternals.MAX_DB_RETRY_COUNT) {
+        try {
+          /*
+           * when DB connection could not be fetched here, dbcp library will keep searching until a timeout defined in
+           * its code hardly.
+           */
+          connection = createDataSource().getConnection();
```

Comment:
I am not too concerned as long as there is an upper bound on the threads from AZ hosts that will call this method.